src/backend/distributed/commands/function.c

```diff
@@ -665,6 +717,102 @@ ShouldPropagateAlterFunction(const ObjectAddress *address)
 }
 
 
+/*
+ * PlanCreateFunctionStmt is called during the planning phase for CREATE [OR REPLACE]
+ * FUNCTION. We primarily care for the replace variant of this statement to keep
+ * distributed functions in sync. We bail via a check on ShouldPropagateCreateFunction
+ * which checks for the OR REPLACE modifier.
+ *
+ * Since we use pg_get_functiondef to get the ddl command we actually do not do any
+ * planning here, instead we defer the plan creation to the processing step.
+ *
+ * Instead we do our basic housekeeping where we make sure we are on the coordinator and
+ * can propagate the function in sequential mode.
+ */
+List *
+PlanCreateFunctionStmt(CreateFunctionStmt *stmt, const char *queryString)
+{
+	if (!ShouldPropagateCreateFunction(stmt))
+	{
+		return NIL;
+	}
+
+	EnsureCoordinator();
+
+	EnsureSequentialModeForFunctionDDL();
+
+	/*
+	 * ddl jobs will be generated during the Processing phase as we need the function to
+	 * be updated in the catalog to get its sql representation
+	 */
+	return NIL;
+}
+
+
+/*
+ * ProcessCreateFunctionStmt actually creates the plan we need to execute for function
+ * propagation. This is the downside of using pg_get_functiondef to get the sql statement.
+ *
+ * Besides creating the plan we also make sure all (new) dependencies of the function are
+ * created on all nodes.
+ */
+List *
+ProcessCreateFunctionStmt(CreateFunctionStmt *stmt, const char *queryString)
+{
+	const ObjectAddress *address = NULL;
+	const char *sql = NULL;
+	List *commands = NIL;
+
+	if (!ShouldPropagateCreateFunction(stmt))
+	{
+		return NIL;
+	}
+
+	address = GetObjectAddressFromParseTree((Node *) stmt, false);
+	EnsureDependenciesExistsOnAllNodes(address);
+
+	sql = GetFunctionDDLCommand(address->objectId);
```

Comment:
I like this approach so much that, I couldn't remember the reason that we have all the logic for function deparsing. Can't we always use this approach for all function related things? We don't have to do that, but I think if we can do it, we'd be able to remove a lot of code. I wouldn't mind sending `CREATE OR REPLACE FUNCTION` to the workers on `ALTER FUNCTION` on the coordinator. Are there any specific reasons/concerns for not doing it?

Reply:
I think the main uses of ALTER FUNCTION (ownership, rename, schema change) probably cannot be done via CREATE OR REPLACE. It could make sense for SET, but not sure if CREATE takes more aggressive locks.

Reply:
Marco's response for ownership, rename and schema changes are the exact reasons.

Looking at the actions base syntax (which includes) `SET` I think we could replace that one with this version. However due to time constraints now I think we should keep our current version and refactor it at a later stage when time is not a limiting factor.